src/connection/security/scram/scram-authentication.ts

```diff
@@ -0,0 +1,124 @@
+import { type ScramCredentials } from './scram-credentials';
+import { type ScramMechanism } from './scram-mechanism';
+import { type ServerFirstMessage } from './server-first-message';
+import { createHash, createHmac, pbkdf2Sync, randomBytes } from 'crypto';
+
+export class ScramAuthentication {
+  private static readonly GS2_HEADER = 'n,,';
+  private readonly clientNonce: string;
+
+  constructor(
+    private readonly credentials: ScramCredentials,
+    private readonly mechanism: ScramMechanism
+  ) {
+    this.clientNonce = this.newNonce();
+  }
+
+  /**
+   * @see https://datatracker.ietf.org/doc/html/rfc5802#section-7
+   */
+  public getClientFirstMessage(): string {
+    return `${ScramAuthentication.GS2_HEADER}${this.getClientFirstMessageBare()}`;
+  }
+
+  /**
+   * @see https://datatracker.ietf.org/doc/html/rfc5802#section-3
+   * @see https://datatracker.ietf.org/doc/html/rfc5802#section-7
+   */
+  public getClientFinalMessage(serverFirstMessage: ServerFirstMessage): string {
+    if (!serverFirstMessage.serverNonce.startsWith(this.clientNonce)) {
+      throw new Error('Invalid server nonce: does not start with client nonce');
+    }
+    //ClientKey = HMAC(SaltedPassword, "Client Key")
+    const clientKey = createHmac(this.mechanism.name, this.getSaltedPassword(serverFirstMessage))
+      .update('Client Key')
+      .digest();
+
+    //StoredKey = H(ClientKey)
+    const storedKey = createHash(this.mechanism.name).update(clientKey).digest();
+
+    //AuthMessage = client-first-message-bare + "," server-first-message + "," client-final-message-without-proof
+    const clientFirstMessageBare = this.getClientFirstMessageBare();
+    const clientFinalMessageWithoutProof = this.getClientFinalMessageWithoutProof(serverFirstMessage);
+    const authMessage = `${clientFirstMessageBare},${serverFirstMessage.original},${clientFinalMessageWithoutProof}`;
+
+    //ClientSignature = HMAC(StoredKey, AuthMessage)
+    const clientSignature = createHmac(this.mechanism.name, storedKey).update(authMessage).digest();
+
+    //ClientProof = ClientKey XOR ClientSignature
+    const clientProof = this.xor(clientKey, clientSignature).toString('base64');
+
+    return `${clientFinalMessageWithoutProof},p=${clientProof}`;
+  }
+
+  /**
+   * @see https://datatracker.ietf.org/doc/html/rfc5802#section-7
+   */
+  private getClientFinalMessageWithoutProof(serverFirstMessage: ServerFirstMessage): string {
+    const channelBinding = `c=${Buffer.from(ScramAuthentication.GS2_HEADER).toString('base64')}`;
+
+    return `${channelBinding},r=${serverFirstMessage.serverNonce}`;
+  }
+
+  private xor(bufferA: Buffer, bufferB: Buffer): Buffer {
+    if (Buffer.byteLength(bufferA) !== Buffer.byteLength(bufferB)) {
+      throw new Error('Argument arrays must be of the same length');
+    }
+
+    const result = [];
```

Comment:
Consider preallocating Buffer and writing directly to each index
```suggestion
    const result = Buffer.alloc(bufferA.length);
```